### DISCUSSION TITLE
Importing CreateCustomContainerTrainingJobOperator takes too much time

### DISCUSSION THREAD

**Discussion by shahar1** [apache/airflow#29615] [General] [answered]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### What happened

When importing CreateCustomContainerTrainingJobOperator it seems to take 3-4 seconds just to import the module.
I ran `time python` for two DAG files - one with the import line and one without it. The import line is simple as follows:
``from airflow.providers.google.cloud.operators.vertex_ai.custom_job import CreateCustomContainerTrainingJobOperator``

The results of the comparison:

```
airflow@airflow-scheduler-##########-#####:~/gcs/dags$ time python dag_w_import.py
real    0m4.884s
user    0m4.668s
sys     0m1.425s
airflow@airflow-scheduler-##########-#####:~/gcs/dags$ time python same_dag_wo_import.py
real    0m1.484s
user    0m1.310s
sys     0m0.163s

```

### What you think should happen instead

The import of the module should be instant

### How to reproduce

Described above

### Operating System

Ubuntu 18.04.6 LTS

### Versions of Apache Airflow Providers

apache-airflow-providers-google==8.6.0

### Deployment

Google Cloud Composer

### Deployment details

composer-2.1.6-airflow-2.4.3

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by Taragolis** [comment, marked as answer]
Check it locally, most of the time spend during this import:

`from google.cloud.aiplatform.models import Model`

![image](https://user-images.githubusercontent.com/3998685/219967078-195efa4c-e9d3-4ba9-a601-a032dea4accc.png)

**Reply by potiuk** [reply]
Yeah. I am not sure where your idea is that imports should work "instantly" @shahar1 - this is a fact of life that some imports take longer. You can raise an issue in `google.cloud.airplatform` package if you think the import time should be somehow shortened. We can do literally nothing to help with it.

**Reply by Taragolis** [reply]
I think we already partially done the same thing for `pandas` which import might take a lot of time.
But with pandas it is quite simple, because it is not something new and we know `pandas` and `numpy` might take significant time for import.

But we don't know anything about other third parties libraries. Potentially the solution might be the same:
1. Move `from google.cloud.aiplatform.models import Model` under `if TYPE_CHECKING:` block for static checks
2. Use local import in `execute`
3. And wait for [PEP 690 ](https://peps.python.org/pep-0690/), 5+ years to be honest, first need to wait Python 3.12 release and after that until it would be oldest supported interpreter

Another option: address this issue to https://github.com/googleapis/python-aiplatform/issues because long import happen in this library

**Reply by potiuk** [reply]
> Another option: address this issue to https://github.com/googleapis/python-aiplatform/issues because long import happen in this library

This. Maybe @shahar1 you can create an issue there?

**Reply by shahar1** [reply]
> > Another option: address this issue to https://github.com/googleapis/python-aiplatform/issues because long import happen in this library
> 
> This. Maybe @shahar1 you can create an issue there?

Yes, I will do it soon. Thanks!
Question about @Taragolis 's first suggestion - why is the 3rd step (PEP 690) necessary? Isn't it enough for now to just move the local import into the `execute()` function?

**Reply by potiuk** [reply]
> Question about @Taragolis 's first suggestion - why is the 3rd step (PEP 690) necessary? Isn't it enough for now to just move the local import into the execute() function?

Try it - you will see. Lots of code duplication, lots of if TYPE_CHECKING. PEP 690 is supposed to avoid all that.

**Reply by shahar1** [reply]
Opened an issue at [python-aiplatform](https://github.com/googleapis/python-aiplatform/issues/1994) repository. Hopefully, it could be resolved at their level.
